2009-07-02  Ulrich Drepper  <drepper@redhat.com>

	* malloc/malloc.c [ATOMIC_FASTBINS] (_int_free): Add full barrier when
	adding to fast bin list.

--- libc/malloc/malloc.c
+++ libc/malloc/malloc.c
@@ -4822,6 +4822,7 @@ _int_free(mstate av, mchunkptr p)
 	    goto errout;
 	  }
 	p->fd = fd = old;
+	atomic_full_barrier ();
       }
     while ((old = catomic_compare_and_exchange_val_acq (fb, p, fd)) != fd);
 #else
